### DISCUSSION TITLE
Cross-compilation for Nelua?

### DISCUSSION THREAD

**Discussion by rprata** [edubart/nelua-lang#37] [Q&A] [answered]
Hi, 

Is it possible to add support for cross-compilation? For example, we can set `toolchain`, `compiler version`, `rootfs` and other tools? It is a good feature for embedded system.

**Comment by edubart** [comment, marked as answer]
You just need to use a different C compiler or CFLAGS. I do this with `nelua --cc=x86_64-w64-mingw32-gcc helloworld.nelua` for example to cross compile to Windows while on Linux and `nelua --cc=emcc helloworld.nelua` to cross compile to WebAssembly. Usually any platform that supports C is supported.

**Reply by Hyperboid** [reply]
It also respects the `CC` environment variable. Learned this the hard way because I was trying to compile regular lua for Nintendo 64 and forgot to clean up.

**Comment by jrfondren** [comment]
A missing feature for cross-compilation is for nelua to know, at nelua compile-time, before the C compiler gets the result, what the target platform is. This would allow a program to use linux APIs when compiling to Linux, and Windows APIs when compiling to windows.  An example of a library like that is https://github.com/treeform/puppy , which uses completely different underlying HTTP client libraries depending on the platform. As is, tests like `require("nelua/utils/platform").is_windows` return information appropriate for the host and not the target platform.

Of course a workaround is to have the build system inject globals:
```lua
## if target.os == 'windows' then
print 'on windows'
## elseif target.os == 'linux' then
print 'on linux'
## else static_error('unsupported target') end
```
Compiled as:
```
$ nelua -i '##[[target = { os="windows" }]] require("targeted")'
on windows
$ nelua -i '##[[target = { os="freestanding" }]] require("targeted")'
eval_N85znEeEG1t:1:1: from: AST node Block
##[[target = { os="freestanding" }]] require("targeted")
^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
targeted.nelua:5: error: unsupported target
```

But without a convention, libraries like puppy aren't as convenient

**Reply by jrfondren** [reply]
as soon as I post this I find the feature! This targeted.nelua works:
```lua
## if ccinfo.is_windows then
print 'on windows'
## elseif ccinfo.is_linux then
print 'on linux'
## else static_error('unsupported target') end
```
